### PR TITLE
IAsyncEnumerable on DelayedQuery

### PR DESCRIPTION
**Pull request by snaumenko-st**
Introduce IAsyncEnumerable on DelayedQuery to utilize async Session query while asynchronously prefetching data.

### CODE REVIEW

**Comment by SergeiPavlov** [comment]
Looks like it duplicates 
`DelayedQuery.ExecuteAsync().AsAsyncEnumerable()` functionality

Generally we add a new code into DataObject only when it is impossible or difficult to get the the result by existing DO API.

**Comment by snaumenko-st** [comment]
We need IAsyncEnumerable implementation here as if it is used inside a Prefetch, which is a fairly common case, it is used as the IEnumerable source and will fall back to a synchronous implementation even in a chain of asynchronous calls. The new unit test shows that behavior.

File: Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs [hunk 1]
```diff
@@ -376,6 +382,99 @@ await using (var session = await Domain.OpenSessionAsync())
       }
     }
 
+    private class QueryCounterSessionHandlerMoq : ChainingSessionHandler
+    {
+
+      private int syncCounter;
```
Comment (SergeiPavlov) [inline Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs]: Please add `volatile` keyword

**Comment by SergeiPavlov** [comment]
You are right

File: Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs [hunk 2]
```diff
@@ -20,14 +20,23 @@ namespace Xtensive.Orm
...
     /// <inheritdoc/>
     public IEnumerator<TElement> GetEnumerator() => Materialize<TElement>().GetEnumerator();
 
     /// <inheritdoc/>
     IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
+    /// <inheritdoc/>
+    async IAsyncEnumerator<TElement> IAsyncEnumerable<TElement>.GetAsyncEnumerator(CancellationToken token = default)
+    {
+      var elements = await MaterializeAsync<TElement>(token).ConfigureAwaitFalse();
```
Comment (SergeiPavlov) [inline Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs]: It is better to replace invocation of protected `MaterializeAsync<TElement>(token)` by  public `ExecuteAsync<TElement>(token)`

**Review by SergeiPavlov — APPROVED** [review]

**Comment by SergeiPavlov** [comment]
@snaumenko-st  could you please create a clone of this PR into the Upstream repo https://github.com/dataObjects-net/dataobjects-net ?

File: Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs (hunk 2 above)
Comment (snaumenko-st) [reply Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs]: Ok

File: Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs (hunk 1 above)
Comment (snaumenko-st) [reply Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs]: I think it makes no sense here, as it is only used inside Interlocked.Increment operations. volatile keyword has no added value in such cases. But I can add volatile read in Assert section, however it makes no sense to me either.

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs]: It is for documenting purposes

File: Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs [hunk 3]
```diff
@@ -376,6 +382,99 @@ await using (var session = await Domain.OpenSessionAsync())
...
+        var prop = typeof(Session).GetProperty("Handler", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.GetProperty);
+        var handler = prop.GetValue(session) as SessionHandler;
+        var moq = new QueryCounterSessionHandlerMoq(handler);
+        using (sessionAccessor.ChangeSessionHandler(moq))
+
+        using (var transactionScope = session.OpenTransaction()) {
+          var people = session.Query.CreateDelayedQuery(q => q.All<Person>())
+            .Prefetch(p => p.Photo) // Lazy load field
+            .Prefetch(p => p.Employees // EntitySet Employees
+              .Prefetch(e => e.Photo)) // and lazy load field of each of its items
```
Comment (SergeiPavlov) [inline Orm/Xtensive.Orm.Manual/Prefetch/PrefetchTest.cs:436]: indentation

File: Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs [hunk 4]
```diff
@@ -20,14 +20,23 @@ namespace Xtensive.Orm
...
+  public sealed class DelayedQuery<TElement> : DelayedQuery, IEnumerable<TElement>, IAsyncEnumerable<TElement>
   {
     /// <inheritdoc/>
     public IEnumerator<TElement> GetEnumerator() => Materialize<TElement>().GetEnumerator();
 
     /// <inheritdoc/>
     IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
+    /// <inheritdoc/>
+    async IAsyncEnumerator<TElement> IAsyncEnumerable<TElement>.GetAsyncEnumerator(CancellationToken token = default)
```
Comment (SergeiPavlov) [inline Orm/Xtensive.Orm/Orm/DelayedQuery{T}.cs]: Looks like explicitly defined interface methods should be written without default arg values.

**Pull request merged**